src/dmd/semantic2.d

```diff
@@ -393,35 +395,29 @@ private extern(C++) final class Semantic2Visitor : Visitor
                 if (f1.overrides(f2))
                     return 0;
 
+                auto tf2 = cast(TypeFunction) f2.type;
+
                 // extern (C) functions always conflict each other.
                 if (f1.ident == f2.ident &&
                     f1.toParent2() == f2.toParent2() &&
                     (f1.linkage != LINK.d && f1.linkage != LINK.cpp) &&
-                    (f2.linkage != LINK.d && f2.linkage != LINK.cpp))
+                    (f2.linkage != LINK.d && f2.linkage != LINK.cpp) &&
+
+                    // But allow the hack to declare overloads with different parameters/STC's
+                    (!tf1.attributesEqual(tf2) || tf1.parameterList != tf2.parameterList))
                 {
-                    /* Allow the hack that is actually used in druntime,
-                     * to ignore function attributes for extern (C) functions.
-                     * TODO: Must be reconsidered in the future.
-                     *  BUG: https://issues.dlang.org/show_bug.cgi?id=18206
-                     *
-                     *  extern(C):
-                     *  alias sigfn_t  = void function(int);
-                     *  alias sigfn_t2 = void function(int) nothrow @nogc;
-                     *  sigfn_t  bsd_signal(int sig, sigfn_t  func);
-                     *  sigfn_t2 bsd_signal(int sig, sigfn_t2 func) nothrow @nogc;  // no error
-                     */
-                    if (f1.fbody is null || f2.fbody is null)
-                        return 0;
-
-                    auto tf2 = cast(TypeFunction)f2.type;
-                    error(f2.loc, "%s `%s%s` cannot be overloaded with %s`extern(%s)` function at %s",
+                    // @@@DEPRECATED_2.094@@@
+                    // Deprecated in 2020-08, make this an error in 2.104
+                    deprecation(f2.loc, "%s `%s%s` cannot be overloaded with %s`extern(%s)` function at %s",
```

Comment:
1) Shouldn't we call  `f2.deprecation`  ?
2) It's a weird phrasing: f2 cannot be overloaded by f1 (which is first in lexical order). Isn't it backwards ?
     I would expect something like "<func name> cannot overload `extern (C)` `function` `<FQN>` defined at `%s`"